File: myintent/src/main/java/pl/mareklangiewicz/myintent/PlayStopButton.java
package pl.mareklangiewicz.myintent;

import android.animation.ObjectAnimator;
import android.graphics.drawable.Drawable;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.ImageView;

import com.noveogroup.android.log.MyAndroidLogger;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import pl.mareklangiewicz.mydrawables.MyPlayStopDrawable;
import pl.mareklangiewicz.myutils.IMyLogger;

/**
 * Created by Marek Langiewicz on 15.12.15.
 */
public final class PlayStopButton {

    private @NonNull final IMyLogger log = MyAndroidLogger.UIL;

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({
            HIDDEN,
            PLAY,
            STOP
    })
    public @interface State {
    }

    public static final int HIDDEN = 0;
    public static final int PLAY = 1;
    public static final int STOP = 2;


    private @State int mState = HIDDEN;

    @SuppressWarnings("FieldCanBeLocal")
    private final Drawable mDrawable = new MyPlayStopDrawable()
            .setColorFrom(0xff0000c0)
            .setColorTo(0xffc00000)
            .setRotateTo(90f)
            .setStrokeWidth(6);

    private @NonNull ImageView mView;
    private @NonNull ObjectAnimator mAnimator;

    private @Nullable Listener mListener;

    private final int ANIM_DURATION = 300;

    PlayStopButton(@NonNull ImageView view) {
        mView = view;
        mView.setImageDrawable(mDrawable);
        mAnimator = ObjectAnimator.ofInt(mDrawable, "level", 0, 10000).setDuration(ANIM_DURATION);
        mView.setOnClickListener(new View.OnClickListener() {
            @Override public void onClick(View v) {
                if(mState == HIDDEN) {
                    log.v("Hidden button clicked. ignoring..", null);
                    return;
                }
                setState(mState == STOP ? PLAY : STOP, true, false);
            }
        });
        setState(HIDDEN, false, true);
    }


    @SuppressWarnings("unused")
    public @State int getState() { return mState; }

    public void setState(@State int state) {
        setState(state, false, false);
    }

    private void setState(@State int state, boolean byUser, boolean immediately) {

        @State int oldState = mState;
        mState = state;

        if(immediately) {
            mAnimator.cancel();
            mView.animate().cancel();
            mView.setAlpha(mState == HIDDEN ? 0f : 1f);
            mAnimator.setCurrentPlayTime(mState == PLAY ? 0 : ANIM_DURATION);
        }
        else if(oldState != mState) {
            if(mState == HIDDEN) {
                mView.animate().alpha(0f);
            }
            else if(oldState == HIDDEN) {
                mView.setAlpha(0f);
                mAnimator.cancel();
                if(mState == PLAY) {
                    mAnimator.reverse();
                    mView.animate().alpha(1f);
                }
                else if(mState == STOP) {
                    mAnimator.start();
                    mView.animate().alpha(1f);
                }
            }
            else { // play -> stop or stop -> play
                if(mAnimator.isRunning()) {
                    mAnimator.reverse();
                }
                else { // we are stopped at some end
                    if(oldState == STOP && mState == PLAY) {
                        mAnimator.reverse();
                    }
                    else if(oldState == PLAY && mState == STOP) {
                        mAnimator.start();
                    }
                    else
                        log.a("Incorrect animated button state.", null);
                }
            }

        }

        if(mListener != null)
            mListener.onPlayStopChanged(oldState, mState, byUser);
    }

    @SuppressWarnings("unused")
    public @Nullable Listener getListener() {
        return mListener;
    }

    public void setListener(@Nullable Listener listener) {
        mListener = listener;
    }


    public interface Listener {
        void onPlayStopChanged(@State int oldState, @State int newState, boolean byUser);
    }

}

File: myintent/src/main/java/pl/mareklangiewicz/myintent/PlayStopButton.kt
package pl.mareklangiewicz.myintent

import android.animation.ObjectAnimator
import android.view.View
import android.widget.ImageView
import pl.mareklangiewicz.mydrawables.MyPlayStopDrawable
import pl.mareklangiewicz.myintent.PlayStopButton.State.*
import pl.mareklangiewicz.myloggers.MY_DEFAULT_ANDRO_LOGGER
import pl.mareklangiewicz.myutils.*

/**
 * Created by Marek Langiewicz on 15.12.15.
 */

class PlayStopButton(private val view: ImageView) {

    enum class State {
        HIDDEN, PLAY, STOP
    }

    val ANIM_DURATION = 300L

    private val log = MY_DEFAULT_ANDRO_LOGGER


    private val drawable = MyPlayStopDrawable().apply { colorFrom = 0xff0000c0.toInt(); colorTo = 0xffc00000.toInt(); rotateTo = 90f; strokeWidth = 6f }
    private val animator = ObjectAnimator.ofInt(drawable, "level", 0, 10000).setDuration(ANIM_DURATION)

    var listener: Listener? = null


    init {
        view.setImageDrawable(drawable)
        view.setOnClickListener(View.OnClickListener {
            val oldState = state
            if (oldState == HIDDEN) {
                log.v("Hidden button clicked. ignoring..")
                return@OnClickListener
            }
            state = if (oldState == STOP) PLAY else STOP
            listener?.onPlayStopClicked(oldState, state)
        })
        view.alpha = 0f
        animator.currentPlayTime = ANIM_DURATION
    }

    var state: State = HIDDEN
        set(newState) {
            if (newState == field)
                return
            val oldState = field
            field = newState

            if (newState == HIDDEN) {
                view.animate().alpha(0f)
            } else if (oldState == HIDDEN) {
                view.alpha = 0f
                animator.cancel()
                if (newState == PLAY) {
                    animator.reverse()
                    view.animate().alpha(1f)
                } else if (newState == STOP) {
                    animator.start()
                    view.animate().alpha(1f)
                }
            } else {
                // play -> stop or stop -> play
                if (animator.isRunning) {
                    animator.reverse()
                } else {
                    // we are stopped at some end
                    if (oldState == STOP && newState == PLAY) {
                        animator.reverse()
                    } else if (oldState == PLAY && newState == STOP) {
                        animator.start()
                    } else
                        log.a("Incorrect animated button state.")
                }
            }
        }

    interface Listener { // TODO SOMEDAY: use MyMachines.Relay to publish changes
        fun onPlayStopClicked(oldState: State, newState: State)
    }

}
